util.go

```diff
@@ -0,0 +1,99 @@
+package semver
+
+import "unicode"
+
+func compareSegment(v, o uint64) int {
+	if v < o {
+		return -1
+	}
+	if v > o {
+		return 1
+	}
+	return 0
+}
+
+func isBuildIdentifier(s string) bool {
+	return isDigits(s) || isAlphaNumericIdentifier(s)
+}
+
+func isPreReleaseIdentifier(s string) bool {
+	return isAlphaNumericIdentifier(s) || isNumericIdentifier(s)
+}
+
+func isAlphaNumericIdentifier(s string) bool {
+	if len(s) == 1 && isNonDigit(rune(s[0])) {
+		// must be a non-diget if len==1
+		return false
+	}
+	// must contain one non-diget
+	var foundNonDigit bool
+	for _, char := range s {
+		if isNonDigit(char) {
+			foundNonDigit = true
+		}
+		if !isIdentifierChar(char) {
+			return false
+		}
+	}
+	return foundNonDigit
+}
+
+func isNumericIdentifier(s string) bool {
+	if s == "0" {
+		return true
+	}
+	for _, char := range s {
+		if !isPositiveDigit(char) {
+			// must start with positive digit
+			return false
+		}
```

Comment:
Strictly non-zero positive integers are required for the first digit only. As implemented this enforces that no subsequent digits be `0` precluding versions such as `1.0.0-1.10.0`.

An additional test case in the `version_test.go` file will reveal this:
```go
func TestVersion_String(t *testing.T) {
	t.Parallel()
	// input equals output test
	tests := []string{
		"0.0.4", "10.20.30", "1.1.2-prerelease+meta",
		"1.0.0-alpha.beta.1", "1.0.0-alpha-a.b-c-somethinglong+build.1-aef.1-its-okay",
		"9999999999999999999.9999999999999999999.9999999999999999999",
		"1.2.3----RC-SNAPSHOT.12.9.1--.12+788",
		"1.0.0-1.10.0", //Add this testcase
	}
	for _, test := range tests {
		test := test
		t.Run(test, func(t *testing.T) {
			t.Parallel()
			v, err := NewVersion(test)
			require.NoError(t, err)
			assert.Equal(t, test, v.String())
		})
	}
}
```

Reply:
Wow, nice find!
Pushed an updated commit. New test case is in the versionparser_test.go though. :)